cranelift-sizeless-vector.md

```diff
@@ -18,12 +18,80 @@ Rust is also currently starting to use LLVM's ScalableVectorType and, as a targe
 # Proposal
 [proposal]: #proposal
 
-We can add sizeless vector types by modifying existing structs to hold an extra bit of information to represent the sizeless nature:
+The following proposal includes changes to the type system, adding specific entities for sizeless stack slots as well as specific instructions that take those entities as operands.
 
-- The new types don't report themselves as vectors, so ty.is\_vector() = false, but are explicitly reported via ty.is\_sizeless\_vector().
+We can add sizeless vector types by modifying existing structs to hold an extra bit of information to represent the sizeless nature.
+
+## Type System
+- The new types do not report themselves as vectors, so ty.is\_vector() = false, but are explicitly reported via ty.is\_sizeless\_vector().
+- is\_vector is also renamed to is\_sized\_vector to avoid ambiguity.
 - The TypeSet and ValueTypeSet structs gain a bool to represent whether the type is sizeless.
 - TypeSetBuilder also gains a bool to control the building of those types.
-- At the encoding level, a bit is used to represent whether the type is sizeless, this bit has been taken from the max number of vector lanes supported, so they'd be reduced to 128 from 256.
+- At the encoding level, a bit is used to represent whether the type is sizeless and this bit has been taken from special types range.
+- These changes allow the usual polymorphic vector operations to be automatically built for the new set of sizeless types.
+
+## IR Entities and Function Changes
+A new entity is added to the IR, the SizelessStackSlot, and the Function will hold these separately from the existing StackSlot entities, also providing two APIs to create them:
+- create\_sized\_stack\_slot
+- create\_sizeless\_stack\_slot
+
+Keeping two vectors enables us to continue to use each entity's index as it's slot identifier, and allows us to place the entities in different positions in the frame. It also enables the backends to easily track which slots are sizeless, and which are not.
+
+## Instructions
+Three new instructions are added at the IR level to use the new SizelessStackSlot entity:
+- SizelessStackAddr
+- SizelessStackLoad
+- SizelessStackStore
+
+The primary difference between these operations and their existing counterparts is that they only take a SizelessStackSlot operand, without a byte offset.
+
+## ABI Layer Changes
+
+- The method stack_stackslot_addr is renamed to sized\_stackslot\_addr.
+- The method sizeless\_stackslots\_addr is introduced, which takes a vector\_scale parameter.
```

Comment:
What does "vector_scale" mean in this context? Number of basic-vector-register units in the vector value's width, or...?

Reply:
This is the number of bytes of a  'sizeless' vector, allowing us to take a slot index and calculate an address.